src/components/validation/runner/mod.rs

```diff
@@ -400,6 +388,53 @@ impl Runner {
 
         Ok(test_case_results)
     }
+
+    // TODO I think a nice refactoring of the TaskCoordinator / it's usage in the Runner could be helpful
+    // , to organize all of the task coordinators into a single struct and have methods like `new()` and
+    // `shutdown()`.
+    async fn coordinate_shutdown(
+        &self,
+        input_task_coordinator: &mut TaskCoordinator<Started>,
+        telemetry_task_coordinator: &mut TaskCoordinator<Started>,
+        topology_task_coordinator: &mut TaskCoordinator<Started>,
+        output_task_coordinator: &mut TaskCoordinator<Started>,
+        output_driver: JoinHandle<Vec<Event>>,
+    ) -> Vec<Event> {
+        // for sinks, we need drive the shutdown by ensuring that the output events have been
+        // processed by the external resource, which ensures that the input events have travelled
+        // all the way through the pipeline, and that the telemetry events have been processed
+        // before shutting down the telemetry and topology tasks.
+        if self.configuration.component_type == ComponentType::Sink {
+            input_task_coordinator.shutdown().await;
+
+            output_task_coordinator.shutdown().await;
+
+            let output_events = output_driver
+                .await
+                .expect("output driver task should not have panicked");
+
+            telemetry_task_coordinator.shutdown().await;
+
+            topology_task_coordinator.shutdown().await;
+
+            output_events
+
+        // for sources, and transforms the output events are collected from the controlled edge (vector sink),
+        // and as such the coordination of shutdown of the tasks is more straight forward.
+        } else {
+            input_task_coordinator.shutdown().await;
+
+            telemetry_task_coordinator.shutdown().await;
+
+            topology_task_coordinator.shutdown().await;
+
+            output_task_coordinator.shutdown().await;
+
+            output_driver
+                .await
+                .expect("output driver task should not have panicked")
+        }
+    }
```

Comment:
I might be misreading this, but I wonder: could you not just use the first approach for all cases?

Like, if the first approach is working for the purpose of collecting/waiting long enough for all output events as well as telemetry data... then I'm trying to think of why it also wouldn't work for sources and transforms? 🤔

Reply:
I had this thought as well when I was writing this. I had tried what you suggested and it didn't work for sources. I just tried it again and I realized there is a 3 birds one stone solution to the issues you have raised.

Instead of having the external resource be responsible for reporting if events received equals expected, we can do that in the output driver itself. This means sources will be able to use the same shutdown order, because the controlled edge doesn't need to coordinate things. And it means any other external resource implementation doesn't need to have that logic repeated.

It also presented an opportunity to add a timeout mechanism so that we still gracefully shutdown.